filepathfilter/filepathfilter.go

```diff
@@ -57,311 +49,129 @@ func patternsToStrings(ps ...Pattern) []string {
-			if matched {
-				pattern = inc.String()
-				break
-			}
-		}
-		if !matched {
-			return "", false
-		}
-	}
+type wm struct {
+	w    *wildmatch.Wildmatch
+	p    string
+	dirs bool
+}
 
-	if len(f.exclude) > 0 {
-		for _, ex := range f.exclude {
-			if ex.Match(cleanedName) {
-				return ex.String(), false
-			}
+func (w *wm) Match(filename string) bool {
+	return w.w.Match(w.chomp(filename))
+}
+
+func (w *wm) chomp(filename string) string {
+	for _, suffix := range []string{`/`, `\`} {
+		if strings.HasSuffix(filename, suffix) {
+			return strings.TrimSuffix(filename, suffix)
 		}
 	}
+	return filename
+}
 
-	return pattern, true
+func (w *wm) String() string {
+	return w.p
 }
 
 const (
-	sep = string(filepath.Separator)
+	sep byte = '/'
 )
 
-func NewPattern(rawpattern string) Pattern {
-	cleanpattern := filepath.Clean(rawpattern)
-
-	// Special case local dir, matches all (inc subpaths)
-	if _, local := localDirSet[cleanpattern]; local {
-		return &noOpMatcher{pattern: cleanpattern}
-	}
-
-	hasPathSep := strings.Contains(cleanpattern, sep)
-	ext := filepath.Ext(cleanpattern)
-	plen := len(cleanpattern)
-	if plen > 1 && !hasPathSep && strings.HasPrefix(cleanpattern, "*") && cleanpattern[1:plen] == ext {
-		return &simpleExtPattern{ext: ext}
-	}
-
-	// special case * when there are no path separators
-	// filepath.Match never allows * to match a path separator, which is correct
-	// for gitignore IF the pattern includes a path separator, but not otherwise
-	// So *.txt should match in any subdir, as should test*, but sub/*.txt would
-	// only match directly in the sub dir
-	// Don't need to test cross-platform separators as both cleaned above
-	if !hasPathSep && strings.Contains(cleanpattern, "*") {
-		pattern := regexp.QuoteMeta(cleanpattern)
-		regpattern := fmt.Sprintf("^%s$", strings.Replace(pattern, "\\*", ".*", -1))
-		return &pathlessWildcardPattern{
-			rawPattern: cleanpattern,
-			wildcardRE: regexp.MustCompile(regpattern),
+func NewPattern(p string) Pattern {
+	pp := p
+
+	// Special case: the below patterns match anything according to existing
+	// behavior.
+	switch pp {
+	case `*`, `*.*`, `.`, `./`, `.\`:
+		pp = join("**", "*")
+	}
+
+	dirs := strings.Contains(pp, string(sep))
+	rooted := strings.HasPrefix(pp, string(sep))
+	wild := strings.Contains(pp, "*")
+
+	if !dirs && !wild {
+		// Special case: if pp is a literal string (optionally including
+		// a character class), assume it is a substring match.
+		pp = join("**", pp, "**")
+	} else {
+		if dirs && !rooted {
+			// Special case: if there are any directory separators,
+			// assume "pp" is rooted.
+			if !wild {
+				pp = join("**", pp, "**")
+			}
+		} else {
+			if rooted {
+				// Special case: if there are not any directory
+				// separators, assume "pp" is a substring match.
+				pp = join(pp, "**")
+			} else {
+				// Special case: if there are not any directory
+				// separators, assume "pp" is a substring match.
+				pp = join("**", pp)
+			}
```

Comment:
I think this "guessing" is dangerous. We never know what the real intention of the users are. I would rather print a big warning message for a few versions that the meaning of this pattern changed.

Reply:
Agreed, though I think that my language is not correct here, in particular with my usage of "assume". This translation from old pattern to new pattern is intended to be behavior-preserving, and this is how this pattern worked before. `foo` would match either the file `foo` at the tree root, or any file `foo` further down [[1]](https://github.com/git-lfs/git-lfs/blob/v2.3.4/filepathfilter/filepathfilter_test.go#L42-L49).

I have changed this comment to better convey my intentions in: 2a17ee5c.